File: src/data/courses/dsa/trie/trie-dfs-backtracking/word-search-ii/code/python/word_search_ii.py
# ============================================================
# Solution 1: DFS per word — O(W*M*N*4^L) — search each separately
# ============================================================
class Solution1:
    def findWords(self, board, words):
        m, n = len(board), len(board[0])
        def dfs(word, i, r, c):
            if i == len(word): return True
            if r<0 or r>=m or c<0 or c>=n or board[r][c] != word[i]: return False
            tmp, board[r][c] = board[r][c], '#'
            found = dfs(word,i+1,r+1,c) or dfs(word,i+1,r-1,c) or dfs(word,i+1,r,c+1) or dfs(word,i+1,r,c-1)
            board[r][c] = tmp
            return found
        result = []
        for w in words:
            for r in range(m):
                for c in range(n):
                    if dfs(w, 0, r, c): result.append(w); break
                else: continue
                break
        return result

# ============================================================
# Solution 2: Trie + single DFS — O(M*N*4^L) — search all at once
# ============================================================
class TrieNode:
    def __init__(self): self.children, self.word = {}, None

class Solution2:
    def findWords(self, board, words):
        root = TrieNode()
        for w in words:
            node = root
            for c in w:
                if c not in node.children: node.children[c] = TrieNode()
                node = node.children[c]
            node.word = w
        m, n, result = len(board), len(board[0]), []
        def dfs(r, c, node):
            if r<0 or r>=m or c<0 or c>=n or board[r][c] == '#': return
            ch = board[r][c]
            nxt = node.children.get(ch)
            if not nxt: return
            if nxt.word: result.append(nxt.word); nxt.word = None
            board[r][c] = '#'
            dfs(r+1,c,nxt); dfs(r-1,c,nxt); dfs(r,c+1,nxt); dfs(r,c-1,nxt)
            board[r][c] = ch
        for r in range(m):
            for c in range(n):
                dfs(r, c, root)
        return result

File: src/data/courses/dsa/trie/trie-dfs-backtracking/word-search-ii/code/cpp/WordSearchIi.cpp
// ============================================================
// Solution 1: DFS per word — search each word independently
// Time: O(W × M × N × 4^L) | Space: O(L)
// ============================================================
#include <vector>
#include <string>
using namespace std;

class Solution1 {
    int m, n;
    bool dfs(vector<vector<char>>& board, string& word, int i, int r, int c) {
        if (i == (int)word.size()) return true;
        if (r<0||r>=m||c<0||c>=n||board[r][c]!=word[i]) return false;
        char tmp = board[r][c]; board[r][c] = '#';
        bool found = dfs(board,word,i+1,r+1,c) || dfs(board,word,i+1,r-1,c)
                  || dfs(board,word,i+1,r,c+1) || dfs(board,word,i+1,r,c-1);
        board[r][c] = tmp;
        return found;
    }
public:
    vector<string> findWords(vector<vector<char>>& board, vector<string>& words) {
        m = board.size(); n = board[0].size();
        vector<string> result;
        for (auto& w : words) { // Search EACH word separately — slow!
            bool found = false;
            for (int r = 0; r < m && !found; r++)
                for (int c = 0; c < n && !found; c++)
                    if (dfs(board, w, 0, r, c)) found = true;
            if (found) result.push_back(w);
        }
        return result;
    }
};

// ============================================================
// Solution 2: Trie + single DFS — search all words at once
// Time: O(M × N × 4^L) | Space: O(W × L)
// ============================================================
class Solution2 {
    struct TrieNode {
        TrieNode* ch[26] = {};
        string* word = nullptr; // Store word at end node
    };
    int m, n;
    vector<string> result;
    
    void dfs(vector<vector<char>>& board, int r, int c, TrieNode* node) {
        if (r<0||r>=m||c<0||c>=n||board[r][c]=='#') return;
        char ch = board[r][c];
        TrieNode* next = node->ch[ch-'a'];
        if (!next) return; // No word in trie matches — PRUNE!
        if (next->word) { result.push_back(*next->word); next->word = nullptr; } // Found!
        board[r][c] = '#'; // Mark visited
        dfs(board,r+1,c,next); dfs(board,r-1,c,next);
        dfs(board,r,c+1,next); dfs(board,r,c-1,next);
        board[r][c] = ch; // Unmark
    }
public:
    vector<string> findWords(vector<vector<char>>& board, vector<string>& words) {
        TrieNode* root = new TrieNode();
        // Build trie from all words
        for (auto& w : words) {
            TrieNode* n = root;
            for (char c : w) {
                if (!n->ch[c-'a']) n->ch[c-'a'] = new TrieNode();
                n = n->ch[c-'a'];
            }
            n->word = &w;
        }
        m = board.size(); n = board[0].size();
        // Single DFS guided by trie
        for (int r = 0; r < m; r++)
            for (int c = 0; c < n; c++)
                dfs(board, r, c, root);
        return result;
    }
};
